challenges/2021-08-31-zero-point/solutions/rust/ericwburden/ericwburden.rs

```diff
@@ -0,0 +1,249 @@
+//------------------------------------------------------------------------------------- 
+// Here we set up the LinkedList struct and parts
+//------------------------------------------------------------------------------------- 
+#[derive(Debug)]
+pub struct Node<T> {
+    value: T,
+    next: Link<T>,
+}
+
+type Link<T> = Option<Box<Node<T>>>;
+
+#[derive(Debug)]
+pub struct LinkedList<T> {
+    head: Link<T>,
+}
+
+impl<T> LinkedList<T> {
+    pub fn new() -> Self {
+        LinkedList { head: None }
+    }
+
+    pub fn from(values: Vec<T>) -> Self {
+        let mut out = Self::new();
+        for value in values {
+            out.push_value(value);
+        }
+        out
+    }
+    
+    pub fn iter<'a>(&'a self) -> LinkedListIter<'a, T> {
+        LinkedListIter { next: self.head.as_deref() }
+    }
+}
+
+// Don't fret too much over this one, it's memory management and not that interesting
+impl<T> Drop for LinkedList<T> {
+    fn drop(&mut self) {
+        let mut cur_link = self.head.take();
+        while let Some(mut boxed_node) = cur_link {
+            cur_link = boxed_node.next.take();
+        }
+    }
+}
+
+
+//------------------------------------------------------------------------------------- 
+// Here we create a PushValue trait and implement it for the LinkedList and Links
+//------------------------------------------------------------------------------------- 
+pub trait PushValue<T> {
+    //! This trait allows for pushing values onto the end of the Linked List
+    fn push_value(&mut self, value: T);
+}
+
+/// The `Link`s do all the heavy lifting. If a `Link` is `None`, then it replaces
+/// itself with a Link containing the pushed value. If the `Link` is already populated,
+/// then it pushes the value to it's `next` Link.
+impl<T> PushValue<T> for Link<T> {
+    fn push_value(&mut self, value: T) {
+        if let Some(n) = self {
+            n.next.push_value(value);
+        } else {
+            let new_node = Box::new(Node { value, next: None });
+            *self = Some(new_node);
+        }
+    }
+}
+
+/// The LinkedList struct just passes the value to its `head`, and thence on down the
+/// chain of Links. If 'head' is missing, use the value as a new `head`.
+impl<T> PushValue<T> for LinkedList<T> {
+    fn push_value(&mut self, value: T) {
+        if self.head.is_none() {
+            let new_node = Box::new(Node { value, next: None });
+            self.head = Some(new_node);
```

Comment:
I keep seeing `Some()`, is this just a way to declare that this has a value and is not none?

Reply:
Pretty much. In Rust, there is no such thing as NULL or None. Instead, there is an enum called `Option` that can either be contain `Some(T)` or `None`, and any time you have an `Option`, the compiler makes you write code that can handle either case.

Reply:
So, if you have a function that you might not want to return a value from (say if you're looking for the first occurrence of a value in a list that may not have that value), you need to return `Option<T>` to let the compiler know you might get a value or you might not.